Cannot resolve values of anonymous enum defined in another function return type

Given this program:
```c
enum {A, B, C} foo() {
  return A;
}

void bar() {
  if (foo() == A) {
  }
}

int main() {
  return 0;
}
```
CIL fails to parse it:
```
foo.c:6: Error: Cannot resolve variable A.
Error: There were errors during merging

Fatal error: exception GoblintCil__Errormsg.Error
```
GCC is completely fine with it.

This is probably the problem in https://github.com/goblint/bench/issues/39 and https://github.com/goblint/bench/issues/41.

--- a/src/frontc/cabs2cil.ml
+++ b/src/frontc/cabs2cil.ml
@@ -6018,24 +6018,25 @@ and doDecl (isglobal: bool) : A.definition -> chunk = function
                };
 	    !currentFunctionFDEC.svar.vdecl <- funloc;
 
-            (* Setup the environment. Add the formals to the locals. Maybe
-              they need alpha-conv  *)
-            enterScope ();  (* Start the scope *)
-
-            IH.clear varSizeArrays;
-
-            (* Enter all the function's labels into the alpha conversion table *)
-            ignore (V.visitCabsBlock (new registerLabelsVisitor) body);
-            currentLoc := funloc; (* registerLabelsVisitor changes currentLoc, so reset it *)
-            (* Visitor doesn't use and touch currentExpLoc, so no need to reset. *)
-
-            (* Do not process transparent unions in function definitions.
-              We'll do it later *)
-            transparentUnionArgs := [];
-
             (* Fix the NAME and the STORAGE *)
             let _ =
               let bt,sto,inl,attrs = doSpecList n specs in
+
+              (* Setup the environment. Add the formals to the locals. Maybe
+                they need alpha-conv  *)
+              enterScope ();  (* Start the scope *)
+              (* Scope must be started after doSpecList, which adds enum values declared in return type to outer scope! *)
+
+              IH.clear varSizeArrays;
+
+              (* Enter all the function's labels into the alpha conversion table *)
+              ignore (V.visitCabsBlock (new registerLabelsVisitor) body);
+              currentLoc := funloc; (* registerLabelsVisitor changes currentLoc, so reset it *)
+              (* Visitor doesn't use and touch currentExpLoc, so no need to reset. *)
+
+              (* Do not process transparent unions in function definitions.
+                We'll do it later *)
+              transparentUnionArgs := [];
               !currentFunctionFDEC.svar.vinline <- inl;
 
               let ftyp, funattr =

--- a/src/frontc/cabs2cil.ml
+++ b/src/frontc/cabs2cil.ml
@@ -1211,10 +1211,6 @@ class registerLabelsVisitor = object
     V.DoChildren
 end
 
-(* Maps local variables that are variable sized arrays to the expression that
-   denotes their length *)
-let varSizeArrays : exp IH.t = IH.create 17
-
 (**** EXP actions ***)
 type expAction =
     ADrop                               (* Drop the result. Only the
@@ -3738,12 +3734,6 @@ and doExp (asconst: bool)   (* This expression is used as a constant *)
                    mean the pointer to the start of the string *)
           | Const(CStr _) -> SizeOf (charPtrType)
 
-                (* Maybe we are taking the sizeof a variable-sized array *)
-          | Lval (Var vi, NoOffset) -> begin
-              try
-                IH.find varSizeArrays vi.vid
-              with Not_found -> SizeOfE e'
-          end
           | _ -> SizeOfE e'
         in
         finishExp empty size !typeOfSizeOf
@@ -6026,8 +6016,6 @@ and doDecl (isglobal: bool) : A.definition -> chunk = function
                 they need alpha-conv  *)
               enterScope ();  (* Start the scope *)
               (* Scope must be started after doSpecList, which adds enum values declared in return type to outer scope! *)
-
-              IH.clear varSizeArrays;
 
               (* Enter all the function's labels into the alpha conversion table *)
               ignore (V.visitCabsBlock (new registerLabelsVisitor) body);

--- a/test/small1/enum-scope.c
+++ b/test/small1/enum-scope.c
@@ -1,0 +1,12 @@
+enum {A, B, C} foo() {
+  return A;
+}
+
+void bar() {
+  if (foo() == A) {
+  }
+}
+
+int main() {
+  return 0;
+}

--- a/test/testcil.pl
+++ b/test/testcil.pl
@@ -254,6 +254,7 @@ addTest("testrun/const16 ");
 addTest("test/deref _GNUCC=1");
 addTest("test_i/empty");
 addTest("test/enum");
+addTest("test/enum-scope");
 addTest("testrun/enum2");
 addTest("test/func");
 addTest("test/funcarg ");

